SI-6277 decorates runtime symbols with auto-init

xeno-by opened the pull request.

xeno-by commented:
review @paulp

also review @gkossakowski, since we need a performance assessment of the change I introduced

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/368/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1076/

jsuereth commented:
Wow, the cleanups to the tests look great!

This gives us a place to override for initialization of symbols, which is good.   My question is how done know which methods have to be overridden or enforce new methods are appropriately overridden?   We may want to look into trait/cake hierarchy design tools to help here.   It's been a huge cause of bugs in views when inherited behavior is changed and people are unaware of classes like this.

That said, I can't think of any better solution to the problem right now.   This lgtm.

 Since greg is on vacation,  I'll push this through community build performance metrics and we can see what it tells us.

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/368/

xeno-by commented:
Yeah currently we can't protect against people forgetting to override stuff. That was discussed before when Martin first introduced decorators (called SynchronizedSymbols back then), and we decided to drop this issue for now.

Maybe we can do something with macro annotations later on to remedy the situation.

Also I've just noticed a bunch of failed tests. Will take a look by the end of the day.

jsuereth commented:
Not a bad idea.  We need to think through a solution, because maintenance burden is high.

 Again, the patch lgtm.   Nice work.

retronym commented:
A code generator for the boilerplate would be very welcome, even it it has to be run manually like genprod.

jsuereth commented:
@retronym +1

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1076/

xeno-by commented:
KITTEH PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/400/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1107/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1107/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/400/

gkossakowski commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -56,6 +56,51 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
   def newFreeTypeSymbol(name: TypeName, info: Type, value: => Any, flags: Long = 0L, origin: String): FreeTypeSymbol =
     new FreeTypeSymbol(name, value, origin) initFlags flags setInfo info
 
+  protected def createAbstractTypeSymbol(owner: Symbol, name: TypeName, pos: Position, newFlags: Long): AbstractTypeSymbol =
```
Why do we need to take owner explicitly instead of using `this` as in old version? It seems like you are passing `this` every time you call any of those methods.

xeno-by replied:
We need to move these methods away from `Symbol`, so that we can reliably override them only in reflective universe. Therefore we have to explicitly pass `this`.

gkossakowski commented on src/reflect/scala/reflect/runtime/DecoratedSymbols.scala:
```diff
@@ -0,0 +1,209 @@
+import scala.tools.nsc.io.AbstractFile
+import scala.reflect.internal.Flags._
+
+trait DecoratedSymbols extends internal.Symbols { thisUniverse: SymbolTable =>
```
Missing comment what's the purpose of `DecoratedSymbols`.

gkossakowski commented on test/files/run/reify_ann1a.scala:
```diff
@@ -21,7 +21,6 @@ object Test extends App {
   // test 2: import and typecheck
   val toolbox = cm.mkToolBox()
   val ttree = toolbox.typeCheck(tree)
-  ttree.foreach(sub => if (sub.hasSymbol) sub.symbol.typeSignature)
```
Why this line has been removed? Why there's no impact on test output?

xeno-by replied:
Printers use `sym.annotations` to print infos about annotations. 

Previously this method didn't perform automatic initialization, hence if no explicit calls to `initialize` are performed (or equivalents such as `info` or `typeSignature`), then no annotation infos were printed.

With this patch `annotations` automatically performs initialization, hence there's no longer need in calling `typeSignature`.

gkossakowski commented:
@xeno-by Those test failures seem legit, right?

gkossakowski commented:
Also, when it comes to performance implications I'm not ready to help on that yet. It will come one day but this will be some day in October I think.

xeno-by commented:
Yes, these are legit. I'm working on it.

paulp commented:
Using "Decorated" as a term of distinction from regular symbols is equivalent to "Different" or "Other" in terms of explanatory power.  They are being decorated for some reason, right? There is some unifying theme to the decorations? That is the word to use.  ReflectedSymbols, I guess, or something like that.

xeno-by committed:
magic classes are no longer entered twice

Previously magic classes such as NothingClass were entered twice
into root mirrors of java universes. No idea how this managed not to blow up.
Now this problem is fixed.

Also, as usual when I gain understanding of something, this allows to come up
with better names. `magicSymbols` are now split into three collections:
`magicallyEnteredClasses`, `magicallyEnteredMethods` and `magicallyHijackedSymbols`.

xeno-by committed:
SI-6277 decorates runtime symbols with auto-init

A non-invasive solution that makes runtime reflection symbols
auto-initialize themselves when certain methods are called.
This is done in such a way that compilation (both compile-time
and runtime) is totally unaffected.

The auto-initializing methods are:
  1) `isXXX` tests, except for:
     1.1) `isLocal` (not determined by flags)
     1.2) `isSkolem` (transient, never created during unpickling)
     1.3) `isErroneous` (transient, erroneous stuff doesn't get pickled)
     1.4) `isAbstractType` (created by pickler as such)
     1.5) `isAliasType` (created by pickler as such)
     1.6) `isPrimitive` (not determined by flags)
     1.7) `isNumeric` (not determined by flags)
     1.8) `isPrimaryConstructor` (not determined by flags)
     1.9) `isPackage` (created by pickler as such)
    1.10) `isPackageClass` (created by pickler as such)
    1.11) `isParameter` (created by pickler as such)
    1.12) `isParamAccessor` (created by pickler as such)
  2) `privateWithin`
  3) `annotations`

Previous implementation of synchronized decorators had a fatal flaw.
It was overriding creation methods by mixing in decorator methods
in creation methods. In this model, a symbol gets decorated if its parent
is decorated, all the way to the roots. However roots were never decorated,
because they are declared as final objects.

I evaluated a few options (-Yoverride-objects is experimental, so I dismissed
it right away, delegating root children creation to universe-wide methods does
not involve core changes, but it produces so much boilerplate that I wouldn't
be able to look Paul in the eye afterwards), but the best one appears to be
moving creators from Symbol to SymbolTable.

This introduces another indirection into symbol creation, so I will ask Greg
to test the performance right away, but at a glance symbol creation is already
riddled with indirections, so an extra one shouldn't hurt.

xeno-by commented:
Duh finally. Kitteh, PLS REBUILD ALL!!

xeno-by commented:
@retronym My current problem with codegeneration here is that methods that are decorated are picked manually and from random sources + some of these methods (listed in the last commit) are decorated in a different fashion. In this setting, I'd reckon that a codegenerator might even be harder to support than the manually written boilerplate.

When we have macro annotations, the situation will improve vastly. Most likely we'll be able to just slap @synchronized or @autoinit annotations on relevant methods and be fine. Therefore I suggest we wait till then.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/425/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1133/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1133/

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/425/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1135/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1135/

xeno-by commented:
Yay it works! @paulp could you please review?

paulp commented:
I would like the naming to be improved.

As I mentioned already, "Decorated" is not a meaningful name.  It's like calling a trait "Trait" - "Decorator" is what it is, not what it does.  The decorations being performed all appear to be involve either synchronization or initialization.  There must be a name which captures either that, or the motivation behind that (reflection, whatever.)

I would really prefer not to use "magic" to label anything.  I'm not averse to the occasional unorthodox name, but when one encounters a name like "magicSymbols" it's just aggravating.  Put yourself in the shoes of someone trying to wrap their head around the compiler.  It's baroque enough without help like this.  Words like "magic" and "special" are in the same category as "decorated" in that what they communicate is: "there is something to be said about this, but I stopped looking for the right name."

Let's take the trouble to establish what it is that makes magical symbols magic and why it is that decorated symbols are decorated, and let those answers point us toward better names.

xeno-by commented:
`Decorated` has been changed to `Reflected` in the latest commit 33d43db (does github show it to you?).

Also `magicSymbols` is now just an umbrella name for `magicallyEnteredClasses ++ magicallyEnteredMethods ++ magicallyHijackedSymbols`, all of which seem to be understandable individually. What do you think?

retronym commented:
> In this setting, I'd reckon that a codegenerator might even be harder to support than the manually written boilerplate.

I don't quite buy it. But how about a middle ground: In the test, reflect on the decorator and the decoratee and assert that all methods are overridden (other than the handful that don't need it). This will guard against us adding a method and forgetting to decorate it.

jsuereth commented:
Cool idea, I like it.  I may steal if for the collection pastry factory.

xeno-by commented:
Actually Martin suggested something that's even better. All those `isXXX` overrides can be reduced to overriding just `hasFlag` and having a mask of flags that should auto-initialize. I'll try to take a look tomorrow.

xeno-by commented:
Oops that won't work, because `hasFlag` is final :(

xeno-by commented:
@retronym accommodated your feedback

paulp commented:
Unfinalize it.

xeno-by commented:
I believe then it won't be inlined.

paulp commented:
On Fri, Sep 14, 2012 at 6:29 AM, Eugene Burmako notifications@github.comwrote:

> I believe then it won't be inlined.
> 
> Whatever it is you mean by that, you believe incorrectly.  I'll explain
> final behavior elsewhere, just trust me and unfinalize it.

paulp commented:
"Also magicSymbols is now just an umbrella name for magicallyEnteredClasses ++ magicallyEnteredMethods ++ magicallyHijackedSymbols, all of which seem to be understandable individually."

The goal is not to be understandable, the goal is to be obvious - and to use standard terminology where standard terminology applies.  I don't like the word "magic" at all, regardless of whether it can be understood if pressed.  It is not magic - "everything happens for a reason" - let's encode the reason in the name, not the handwaving.

xeno-by commented:
Thanks everyone! I'll take your feedback into account.

For now I'm closing this pull request and will reopen it after Greg's performance testing.
